File: server/common/protocol.py
import logging
from common.utils import Bet
from common.protocol_message import ProtocolMessage


class Protocol:
    """Protocol for handling communication between client and server"""

    END_MESSAGE = ProtocolMessage.END_MESSAGE
    BUFFER_SIZE = 4

    @staticmethod
    def send_message(sock, data: bytes) -> int:
        """Send message"""
        data_qty_sent = 0
        data_qty = len(data)

        while data_qty_sent < data_qty:
            sent = sock.send(data[data_qty_sent:])
            if sent == 0:
                raise RuntimeError("Socket connection failed while sending message")
            data_qty_sent += sent

        return data_qty_sent

    @staticmethod
    def receive_message(sock) -> bytes:
        """Receive message until end is found"""
        data = b""
        while True:
            chunk = sock.recv(Protocol.BUFFER_SIZE)
            if not chunk:
                raise RuntimeError(
                    "Socket connection failed before receiving complete message"
                )

            data += chunk

            if Protocol.END_MESSAGE in chunk:
                break

        return data.rstrip(Protocol.END_MESSAGE)

File: client/common/protocol.go
package common

import (
	"bytes"
	"io"
	"net"
)

const (
	BUFFER_SIZE = 4
	END_MESSAGE = "\n"
)

type Protocol struct {
	conn net.Conn
}

func NewProtocol(conn net.Conn) *Protocol {
	return &Protocol{
		conn: conn,
	}
}

func (p *Protocol) SendMessage(data []byte) error {
	dataQtySent := 0
	dataQty := len(data)

	for dataQtySent < dataQty {
		sent, err := p.conn.Write(data[dataQtySent:])
		if err != nil {
			return err
		}
		dataQtySent += sent
	}
	return nil
}

func (p *Protocol) ReceiveMessage() (string, error) {
	data := make([]byte, 0)
	buffer := make([]byte, BUFFER_SIZE)
	end := []byte(END_MESSAGE)

	for {
		n, err := p.conn.Read(buffer)
		if err != nil && err != io.EOF {
			return "", err
		}

		if n == 0 {
			return "", io.EOF
		}

		data = append(data, buffer[:n]...)

		if bytes.Contains(buffer[:n], end) {
			break
		}
	}

	data = bytes.TrimSuffix(data, end)

	return string(data), nil
}
